.github/workflows/release.yml

```diff
@@ -150,38 +193,56 @@ jobs:
 
       - name: Check if version updated
         id: check_version_updated
-        if: ${{ steps.reflect_on_pr_content.outputs.create_pull_requests == 'true' }}
         env:
           BOT_TOKEN: ${{ secrets.BOT_TOKEN }}
         run: |
           # check if version file was changed
-          ./ve1/bin/release-checker --version=${{ steps.check_only_version_in_PR_and_authorized.outputs.PR_version }}
+          ./ve1/bin/release-checker \
+              --version=${{ needs.determine-workflow-conditions.outputs.pr-version }}
 
       - name: Sync the repositories
         id: sync_repositories
-        if: ${{ steps.check_version_updated.outputs.release_updated == 'true' }}
+        if: |
+          steps.check_version_updated.outputs.release_updated == 'true'
         env:
           BOT_NAME: ${{ secrets.BOT_NAME }}
           BOT_TOKEN: ${{ secrets.BOT_TOKEN }}
         run: |
-         ./ve1/bin/releaser --version=${{ steps.check_only_version_in_PR_and_authorized.outputs.PR_version }}\
-                             --pr_dir="./pr-branch" \
-                             --development_dir="./dev-repo" \
-                             --charts_dir="./charts-repo" \
-                             --stage_dir="./stage-repo" \
-                             --dev_pr_body="${{ steps.check_only_version_in_PR_and_authorized.outputs.PR_release_body }}" \
-                             --target_branch="${{ github.event.pull_request.base.ref }}" \
-                             --target_repository="${{ github.event.pull_request.base.repo.full_name }}"
+         ./ve1/bin/releaser \
+            --version=${{ needs.determine-workflow-conditions.outputs.pr-version }}\
+            --pr_dir="./pr-branch" \
+            --development_dir="./dev-repo" \
+            --charts_dir="./charts-repo" \
+            --stage_dir="./stage-repo" \
+            --dev_pr_body="${{ needs.determine-workflow-conditions.outputs.pr-body }}" \
+            --target_branch="${{ github.event.pull_request.base.ref }}" \
+            --target_repository="${{ github.event.pull_request.base.repo.full_name }}"
 
+  determine-merge-and-release-conditions:
+    name: Determine merge and release conditions
+    needs:
+      - determine-workflow-conditions
+      - sync-repos
+    outputs:
+      should-merge: ${{ steps.check_merge_and_release.outputs.merge }}
+      release-tag: ${{ steps.check_merge_and_release.outputs.release_tag }}
+      release-body: ${{ steps.check_merge_and_release.outputs.release_body }}
+    runs-on: ubuntu-22.04
+    # cancelled() required because jobs this workflow depends on can skip.
+    if: |
+      !cancelled() &&
+      needs.determine-workflow-conditions.result == 'success' &&
+      ( needs.sync-repos.result == 'success' || needs.sync-repos.result == 'skipped' )
```

Comment:
I'm having trouble understanding the need for `!cancelled()` and `( needs.sync-repos.result == 'success' || needs.sync-repos.result == 'skipped' )`)

> Possible values [of a job result] are success, failure, cancelled, or skipped.

A job that is skipped (i.e. the attached `if` condition evaluates to `false`) will have its result set to `skipped`. This will not prevent subsequent job to run. Here `( needs.sync-repos.result == 'success' || needs.sync-repos.result == 'skipped' )` feels redundant with the default behavior.

Similarly, the job will not run if the workflow was cancelled. In other words, it will run if the workflow was not cancelled. Again, `!cancelled()` feels redundant.

Maybe I'm missing something, what are we trying to achieve here ?

Reply:
This job needs **sync_repos**, which can skip. If it skips, forcing the execution of this job means that `needs.sync_repos.outputs` will be empty, which we want[1]. The calling code already expects the empty values.

> A job that is skipped (i.e. the attached if condition evaluates to false) will have its result set to skipped. This will not prevent subsequent job to run

In my testing, having a result set to skipped **will** prevent a subsequent job from running if it's defined as a dependency via needs. See [this action run](https://github.com/komish/gh-actions-hackery/actions/runs/12124329009) and its accompanying [workflow file](https://github.com/komish/gh-actions-hackery/blob/f10a9db2d91995cf944acec91308d8a68c4f51a2/.github/workflows/skip-behavior.yml). In this case, "three" needs "two" which skipped, so it doesn't run.

Further, even if the Job in question does NOT depend on a skippable job, if it depends on a descendant of a skippable job, and the skippable job skips, it will also not run. See [this action run](https://github.com/komish/gh-actions-hackery/actions/runs/12124141347)  and its accompanying [workflow file](https://github.com/komish/gh-actions-hackery/blob/c25f0833d32d39ae51efad60e8a3fce5602ab6d4/.github/workflows/skip-behavior.yml).  Job "four" only needs "three", which ran, but "four" will not run because "two" skipped.

As I understand it, the default condition for jobs in workflows is `if: success()`. If I add conditions, they will be appended to the default. E.g.:

```yaml
if: |
  success() &&
  'foo' == 'foo' &&
  ( 'this' == 'this' && 'that' == 'that')
```

If I, however, include my own built-in status check from the array of options (e.g. cancelled(), failure()), then it replaces success(). ([ref](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#status-check-functions)).

So, including `!cancelled()` here just allows me to run these in the event the skippable job in their execution path skips. The other conditions make sure that the relevant jobs in scope have valid outcomes expected for this job to run. They may be more informational than practical, to your point - because should either of those fail, this job will not run.

---

[1] This is a carryover from the previous implementation, which allowed for empty values for certain environment variables. Given the size of this PR already, I opted to maintain the same design instead of refactoring.

Reply:
Thanks for the clarification !